File: app/schemas/prompt.py
"""Pydantic schemas for Prompt management."""
from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field


# ===== Prompt Schemas =====

class PromptBase(BaseModel):
    """Base prompt schema."""
    tag: str = Field(..., min_length=1, max_length=128, description="Unique tag/identifier")
    content: str = Field(..., min_length=1, description="Prompt content")
    description: Optional[str] = Field(None, description="Description")


class PromptCreate(PromptBase):
    """Schema for creating a prompt."""
    pass


class PromptUpdate(BaseModel):
    """Schema for updating a prompt."""
    content: Optional[str] = Field(None, min_length=1, description="Updated content")
    description: Optional[str] = Field(None, description="Updated description")
    create_version: bool = Field(True, description="Whether to save this as a new version")
    change_note: Optional[str] = Field(None, description="Note about changes (for version history)")


class PromptResponse(PromptBase):
    """Schema for prompt response."""
    id: str
    current_version: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_orm_with_uuid(cls, obj: Any) -> 'PromptResponse':
        """Create from ORM object with UUID to string conversion."""
        return cls.model_construct(
            id=str(obj.id),
            tag=obj.tag,
            content=obj.content,
            description=obj.description,
            current_version=obj.current_version,
            is_active=obj.is_active,
            created_at=obj.created_at,
            updated_at=obj.updated_at,
        )


class PromptListResponse(BaseModel):
    """Schema for paginated prompt list."""
    items: List[PromptResponse]
    total: int
    page: int
    size: int
    pages: int


# ===== PromptVersion Schemas =====

class PromptVersionResponse(BaseModel):
    """Schema for prompt version response."""
    id: str
    prompt_id: str
    version: int
    content: str
    description: Optional[str]
    change_note: Optional[str]
    created_at: datetime

    @classmethod
    def from_orm_with_uuid(cls, obj: Any) -> 'PromptVersionResponse':
        """Create from ORM object with UUID to string conversion."""
        return cls.model_construct(
            id=str(obj.id),
            prompt_id=str(obj.prompt_id),
            version=obj.version,
            content=obj.content,
            description=obj.description,
            change_note=obj.change_note,
            created_at=obj.created_at,
        )


class PromptVersionListResponse(BaseModel):
    """Schema for prompt version history."""
    items: List[PromptVersionResponse]
    total: int


class PromptDetailResponse(PromptResponse):
    """Detailed prompt response with version history."""
    versions: List[PromptVersionResponse] = []

File: frontend/src/api/prompt.ts
/**
 * Prompt API service with version management.
 */
import request from './request';

// ===== Types =====

export interface Prompt {
    id: string;
    tag: string;
    content: string;
    description: string | null;
    current_version: number;
    is_active: boolean;
    created_at: string;
    updated_at: string;
}

export interface PromptVersion {
    id: string;
    prompt_id: string;
    version: number;
    content: string;
    description: string | null;
    change_note: string | null;
    created_at: string;
}

export interface PromptDetail extends Prompt {
    versions: PromptVersion[];
}

export interface PromptListResponse {
    items: Prompt[];
    total: number;
    page: number;
    size: number;
    pages: number;
}

export interface PromptVersionListResponse {
    items: PromptVersion[];
    total: number;
}

export interface PromptCreateParams {
    tag: string;
    content: string;
    description?: string;
}

export interface PromptUpdateParams {
    content?: string;
    description?: string;
    create_version?: boolean;
    change_note?: string;
}

// ===== API Functions =====

export const promptApi = {
    /**
     * List all prompts with pagination.
     */
    list: async (page = 1, size = 20, search?: string): Promise<PromptListResponse> => {
        const params = new URLSearchParams({ page: String(page), size: String(size) });
        if (search) params.append('search', search);
        return request.get(`/prompts?${params}`);
    },

    /**
     * Get prompt details with version history.
     */
    get: async (id: string): Promise<PromptDetail> => {
        return request.get(`/prompts/${id}`);
    },

    /**
     * Create a new prompt.
     */
    create: async (data: PromptCreateParams): Promise<Prompt> => {
        return request.post('/prompts', data);
    },

    /**
     * Update a prompt.
     */
    update: async (id: string, data: PromptUpdateParams): Promise<Prompt> => {
        return request.patch(`/prompts/${id}`, data);
    },

    /**
     * Delete a prompt (soft delete).
     */
    delete: async (id: string): Promise<void> => {
        return request.delete(`/prompts/${id}`);
    },

    /**
     * Get all versions of a prompt.
     */
    getVersions: async (promptId: string): Promise<PromptVersionListResponse> => {
        return request.get(`/prompts/${promptId}/versions`);
    },

    /**
     * Get a specific version.
     */
    getVersion: async (promptId: string, version: number): Promise<PromptVersion> => {
        return request.get(`/prompts/${promptId}/versions/${version}`);
    },

    /**
     * Rollback to a specific version.
     */
    rollback: async (promptId: string, version: number): Promise<Prompt> => {
        return request.post(`/prompts/${promptId}/rollback/${version}`);
    },
};
